### DISCUSSION TITLE
Concurrent mock expect toHaveBeenCalledTimes

### DISCUSSION THREAD

**Discussion by aaron-scherling-banno** [vitest-dev/vitest#6232] [Q&A] [answered]
Hello! Vitest is amazing, thank you for all the hard work. Truly the best test runner on the market.

Is it possible to have concurrent mock `toHaveBeenCalledTimes`?  I've read another discussions outlining the fact that if the test has state outside of the test, concurrent mode is not an option. However, since I defined the mocks inside of the test, shouldn't that be considered inside the test scope? There is no state in the module under test, but is mocking considered state? I hope the answer is no 🤞, but I have a feeling it's yes since it mutates the imports 

Code:
```javascript
// myFile.ts
import { thirdPartyFunc } from 'thirdParty'

export const funcUnderTest = () => {
    thirdPartyFunc()

    // some other code

    thirdPartyFunc()
}
```

```javascript
// test file
describe('group', () => {
    beforeEach(() => {
        vi.resetModules()
    })

    test.concurrent('test1', async () => {
        const thirdPartyFuncMock = vi.fn().mockImplementation(() => 'hello')
        vi.doMock(import('thirdParty'), () => {
            return {
                thirdPartyFunc: thirdPartyFuncMock
            }
        })

        const mockedModule = await import('./myFile.ts')

        mockedModule.funcUnderTest()

        expect(thirdPartyFuncMock).toHaveBeenCalledTimes(2)
    })

    test.concurrent('test2', async () => {
        const thirdPartyFuncMock = vi.fn().mockImplementation(() => 'hello')
        vi.doMock(import('thirdParty'), () => {
            return {
                thirdPartyFunc: thirdPartyFuncMock
            }
        })

        const mockedModule = await import('./myFile.ts')

        mockedModule.funcUnderTest()

        expect(thirdPartyFuncMock).toHaveBeenCalledTimes(2)
    })
})
```

 I have a feeling that since we are running in parallel, the mocks are overriding tests mid execution. Is that assumption correct?

Potential variation (since concurrency is unpredictable):
1. test1 loads mocks
2. test1 calls `funcUnderTest()`
3. test1 calls first `thirdPartyFunc()` (test1's mocks correctly counts call instance)
4. test2 loads mocks (as a result test1's mocks are overwritten with test2's mocks)
5. test1 calls second `thirdPartyFunc()` (which is now test2's mocks, thus not adding a call count to test1's mock)
6. test2 finishes execution calling `thirdPartyFunc()` twice (adding two more calls to test2's mock)

Final tally:
test1 mock call count - 1
test2 mock call count - 3

**Comment by sheremet-va** [comment, marked as answer]
Yes, `import` is global and not test specific so one mock overrides another

**Reply by aaron-scherling-banno** [reply]
Would we want to add that to the [mock docs](https://vitest.dev/guide/mocking.html)? I don't see that outlined and it wasn't apparent right away. As a new user to Vitest, I turn on `concurrent` mode wanting more speed, but don't know the implications it has with mocking

**Reply by sheremet-va** [reply]
`concurrent` might now give you more speed. It only matters for async tests and if your CI is not powerful enough, it may even lead to performance degradation

I don't think it makes sense to add this to mocks docs - maybe to concurrent docs?
